commerce_coordinator/apps/lms/tasks.py

```diff
@@ -2,21 +2,74 @@
 LMS Celery tasks
 """
 
-from celery import shared_task
+import json
+
+from celery import Task, shared_task
 from celery.utils.log import get_task_logger
 from django.contrib.auth import get_user_model
 from requests import RequestException
 
 from commerce_coordinator.apps.commercetools.catalog_info.constants import TwoUKeys
 from commerce_coordinator.apps.commercetools.clients import CommercetoolsAPIClient
+from commerce_coordinator.apps.commercetools.utils import send_fulfillment_error_email
 from commerce_coordinator.apps.lms.clients import LMSAPIClient
 
 # Use the special Celery logger for our tasks
 logger = get_task_logger(__name__)
 User = get_user_model()
 
 
-@shared_task(bind=True, autoretry_for=(RequestException,), retry_kwargs={'max_retries': 5, 'countdown': 3})
+class CourseEnrollTaskAfterReturn(Task):    # pylint: disable=abstract-method
+    """
+    Base class for fulfill_order_placed_send_enroll_in_course_task
+    """
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        edx_lms_user_id = kwargs.get('edx_lms_user_id')
+        user_email = kwargs.get('user_email')
+        order_number = kwargs.get('order_number')
+        user_first_name = kwargs.get('user_first_name')
+        course_title = kwargs.get('course_title')
+
+        error_message = (
+            json.loads(exc.response.text).get('message', '')
+            if hasattr(exc, 'response') and exc.response is not None
+            else str(exc)
+        )
+
+        logger.error(
+            f"Task {self.name} failed after max retries with error message: {error_message} "
+            f"for user with User Id: {edx_lms_user_id}, Email: {user_email}, "
+            f"Order Number: {order_number}, Course Title: {course_title}"
+        )
+
+        if (
+            self.request.retries >= self.max_retries
```

Comment:
Do we need the retries check here if we know that this function will only be executed after all the retries are exhausted?

Reply:
You are right. This will execute after all retries practically. but I couldn't find any official doc link stating its link with retries. 
So I added this check to ensure the failure and exhaustion of all retries.

Reply:
as per discussion with @shafqatfarhan, I am going to remove this check.